### ISSUE TITLE
Core-Dump in JVM: API call with invalid database connection pointer

### ISSUE THREAD

**Issue by currious** [couchbase/couchbase-lite-core#1054]
We access the couchbase lite core api via the java client https://github.com/couchbase/couchbase-java-client.
We have a test suite in java with multiple test-classes. Each test-class create an couchbase-lite-core instance. After all tests of a test-class are finished, we stop the couchbase lite database. 

If the test-suite with all test-classes is finished, we got sometimes a core-dump. 
We guess that there's an api request even though when the database is already closed.

**Version** 
Linux RedHat UBI 8.1
Installed libs in our linux distribution:
https://packages.couchbase.com/releases/couchbase-lite-java/2.7.0/couchbase-lite-java-2.7.0.zip

Jvm 1.8 openjdk
Java-Libs:
com.couchbase.client:core-io:1.7.2
com.couchbase.client:java-client:2.7.2
com.couchbase.lite:couchbase-lite-java:2.7.1

**Crash**

```
E/CouchbaseLite/DATABASE:SQLite error (code 21): API call with invalid database connection pointer
 E/CouchbaseLite/DATABASE:SQLite error (code 21): misuse at line 118947 of [884b4b7e50]
 E/CouchbaseLite/DATABASE:SQLite error (code 21): API call with invalid database connection pointer
 E/CouchbaseLite/DATABASE:SQLite error (code 21): misuse at line 157025 of [884b4b7e50]
 E/CouchbaseLite/DATABASE:SQLite error (code 21): API call with invalid database connection pointer
 E/CouchbaseLite/DATABASE:SQLite error (code 21): misuse at line 157100 of [884b4b7e50]
 E/CouchbaseLite/DATABASE:bad parameter or other API misuse (21/21)
 W/CouchbaseLite/DATABASE:{?#8} Caught SQLite exception while vacuuming: bad parameter or other API misuse
 #
 # A fatal error has been detected by the Java Runtime Environment:
 #
 #  SIGSEGV (0xb) at pc=0x00007fc7fd66b6bc, pid=25, tid=0x00007fc80b9ff700
 #
 # JRE version: OpenJDK Runtime Environment (8.0_242-b08) (build 1.8.0_242-b08)
 # Java VM: OpenJDK 64-Bit Server VM (25.242-b08 mixed mode linux-amd64 compressed oops)
 # Problematic frame:
 # C  [libLiteCore.so+0x10b6bc]WARNING: RefCounted object at 0x7fc8064048f0 destructed while it had an invalid refCount of -1
 terminate called after throwing an instance of 'std::runtime_error'
   what():  RefCounted object at 0x7fc8064048f0 destructed while it had an invalid refCount of -1
```

**Dump-File**
Are there any couchbase tools to extract information from the core dump? I cannot attach because it's zipped 225MB

**Comment by currious** [comment]
Here's a link to the core-dump: [core-dump](https://saltsolutions-my.sharepoint.com/:u:/g/personal/lamine_bourib_salt-software_de/EfIAowwI0-ZJhagG63HMTl8BAajGWBUVpVFqwxovVJyPiQ?e=UPZi4K)

**Comment by borrrden** [comment]
Are there any more logs that you could provide.  This error message is very general and is probably the result of a very specific set of circumstances about 10 steps long that could be anywhere.  Something is definitely being over-released and so there is invalid memory access.  Is there a Java stack trace or anything?  @bmeike is there any input you can provide?

**Comment by borrrden** [comment]
Also, I don't know what you mean by "access couchbase lite core via the java client".  The java client you mentioned is for communicating with a Couchbase Server cluster.

**Comment by bmeike** [comment]
No... I have no additional information.  There should be a stack trace in that message, but it appears to have been elided.

**Comment by borrrden** [comment]
Any additional info you can provide to address the above @currious ?

**Comment by currious** [comment]
Sorry for the late reply. 
This is our command: **sbt coverage 'testOnly de.my.ScalaUnitTest' coverageReport**
With this error we create a jvm to run a UnitTest. Before we start the unittest, we create an instance of CBLite. 
If the unittest is finished we delete the database and we will remove the directory at disk where the database file was created.

We use the [AbstractDatabase](https://github.com/couchbase/couchbase-lite-java/blob/bd5322704b66acf7a467b0b2a1e5b2cf6fbf0266/lib/src/shared/main/java/com/couchbase/lite/AbstractDatabase.java)-Object to communicate to CBLite. If our tests ends, we call the [AbstractDatabase.close()-Method](https://github.com/couchbase/couchbase-lite-java/blob/bd5322704b66acf7a467b0b2a1e5b2cf6fbf0266/lib/src/shared/main/java/com/couchbase/lite/AbstractDatabase.java#L741-L754) 

I guess through the call of [CDatabase.free](https://github.com/couchbase/couchbase-lite-java/blob/bd5322704b66acf7a467b0b2a1e5b2cf6fbf0266/lib/src/shared/main/java/com/couchbase/lite/internal/core/C4Database.java#L81-L88) this error will be produced. The handle-key is key that identifies
the CBLite and if this value is -1 or some not valid value, it might be possible that memory/disk will be removed that is not allowed.
 ```
public C4Database(
        String path,
        int flags, String storageEngine, int versioning,
        int algorithm, byte[] encryptionKey)
        throws LiteCoreException {
        this.handle = open(path, flags, storageEngine, versioning, algorithm, encryptionKey);
    }
   public void free() {
        if (shouldRetain) { return; }

        free(handle);//here the handle is corrupted and so got the value -1
        handle = 0L;
    }
private static native void free(long db);
```

Log with invalid handle-value error

```
WARNING: RefCounted object at 0x7f535a84e8a0 destructed while it had an invalid refCount of -1
terminate called after throwing an instance of 'std::runtime_error'
  what():  RefCounted object at 0x7f535a84e8a0 destructed while it had an invalid refCount of -1
/usr/bin/bash: line 125:   646 Aborted                 (core dumped) sbt coverage 'testOnly de.my.ScalaUnitTest' coverageReport
ERROR: Job failed: command terminated with exit code 1

or

WARNING: RefCounted object at 0x7f6ed5df7e00 destructed while it had an invalid refCount of -10000000
terminate called after throwing an instance of 'std::runtime_error'
WARNING: RefCounted object at 0x7f6ed5f42a90 destructed while it had an invalid refCount of -1
  what():  RefCounted object at 0x7f6ed5df7e00 destructed while it had an invalid refCount of -10000000
terminate called recursively
```


Does this help you to analyze the problem?

**Comment by bmeike** [comment]
@currious, Thanks for the update.  That's helpful.

It would be far more useful if you were able to give us the output of adb logcat from around the time of the failure.

I've just reviewed the code and it does look to me as if there is a data race if Database.close, Database.delete are called at the same time from multiple threads, or if the Database object is GCed at the same time either of them is called.  If you are sure you are not calling close and delete on different threads, then it is probably the case that the finalizer is being called at about the same time as one of the above.

I believe the race has been fixed in v 2.8
